### PR TITLE
Fix evaluate for LlamaIndex flavor

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/12976?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12976/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12976
```

</p>
</details>

### What changes are proposed in this pull request?

Solve two issues in evaluation with LLamaIndex models.

1. The input data is wrapped in `inputs` column, which caused an error when creating `QueryBundle` object for input.
2. The output from the pyfunc model of LlamaIndex flavor is a single string. However, `mlflow.evaluate()` does not handle that output type ([code](https://github.com/mlflow/mlflow/blob/master/mlflow/models/evaluation/default_evaluator.py#L1472-L1479)) during aggregation.

This PR fixes these two issues to unblock the journey. The first fix is temporary limited to LlamaIndex flavor to reduce regression risk, but should be generalized to other flavors as well. The second fix is less risky so directly applied to the evaluation logic.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

Screenshot 2024-08-20 at 13 39 22

Screenshot 2024-08-20 at 13 39 14

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Fix input/output format bugs in `mlflow.evaluate()` with the LlamaIndex flavor.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/models/utils.py
Comment (B-Step62) [inline mlflow/models/utils.py:1]: note: only cosmetic change here - giving more description for what these functions are

File: mlflow/utils/autologging_utils/__init__.py [hunk 1]
```diff
@@ -89,7 +89,7 @@
 # suppress warnings for certain libraries that are known to have frequent releases.
 _AUTOLOGGING_SUPPORTED_VERSION_WARNING_SUPPRESS_LIST = [
     "langchain",
-    "llama-index",
+    "llama_index",
```
Comment (B-Step62) [inline mlflow/utils/autologging_utils/__init__.py:92]: note: unrelated small fix (to suppress version warning correctly)

File: mlflow/models/utils.py [hunk 2]
```diff
@@ -1655,10 +1653,18 @@ def _convert_llm_ndarray_to_list(data):
     return data
 
 
-def _convert_llm_input_data(data):
-    import pandas as pd
```
[the remark this answers is not in the document]
Comment (BenWilson2) [reply mlflow/models/utils.py:1659]: We don't. Line 18 in this file. Good catch.

File: tests/llama_index/test_llama_index_evaluate.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    with mlflow.start_run():
+        model_info = mlflow.llama_index.log_model(
+            index=single_index,
+            engine_type=engine_type,
+            artifact_path="llama_index",
+        )
+
+    pyfunc_model = mlflow.pyfunc.load_model(model_info.model_uri)
+
+    data = pd.DataFrame(
```
Comment (BenWilson2) [inline tests/llama_index/test_llama_index_evaluate.py:19]: Can we verify with an additional parametrization config that the behavior is the same for a single-row DataFrame as well, just in case there are any positional-based access of the data within the call stack? Pandas does different collections generation (IIRC) based on whether a DF has a single row vs. multiple rows.

**Review by michael-berk — APPROVED** [review]

File: tests/llama_index/test_llama_index_tracer.py [hunk 4]
```diff
@@ -126,7 +126,7 @@ def test_trace_llm_chat(is_async):
         "raw": ANY,
         "delta": None,
         "logprobs": None,
-        "additional_kwargs": {},
+        "additional_kwargs": {"completion_tokens": 12, "prompt_tokens": 9, "total_tokens": 21},
```
Comment (BenWilson2) [inline tests/llama_index/test_llama_index_tracer.py:129]: I guess we need conditional version logic on this one? Maybe only test in the latest version released since this seems like a newly introduced very useful feature that they added :)

**Review by BenWilson2 — APPROVED** [review]
LGTM! Thanks for the fix @B-Step62 :)
